cpp/src/arrow/array-binary-test.cc

```diff
@@ -85,268 +94,132 @@ class TestStringArray : public ::testing::Test {
     ASSERT_OK(BitUtil::BytesToBits(valid_bytes_, default_memory_pool(), &null_bitmap_));
     null_count_ = CountNulls(valid_bytes_);
 
-    strings_ = std::make_shared<StringArray>(length_, offsets_buf_, value_buf_,
-                                             null_bitmap_, null_count_);
-  }
-
- protected:
-  std::vector<int32_t> offsets_;
-  std::vector<char> chars_;
-  std::vector<uint8_t> valid_bytes_;
-
-  std::vector<std::string> expected_;
-
-  std::shared_ptr<Buffer> value_buf_;
-  std::shared_ptr<Buffer> offsets_buf_;
-  std::shared_ptr<Buffer> null_bitmap_;
-
-  int64_t null_count_;
-  int64_t length_;
-
-  std::shared_ptr<StringArray> strings_;
-};
-
-TEST_F(TestStringArray, TestArrayBasics) {
-  ASSERT_EQ(length_, strings_->length());
-  ASSERT_EQ(1, strings_->null_count());
-  ASSERT_OK(ValidateArray(*strings_));
-}
-
-TEST_F(TestStringArray, TestType) {
-  std::shared_ptr<DataType> type = strings_->type();
-
-  ASSERT_EQ(Type::STRING, type->id());
-  ASSERT_EQ(Type::STRING, strings_->type_id());
-}
-
-TEST_F(TestStringArray, TestListFunctions) {
-  int pos = 0;
-  for (size_t i = 0; i < expected_.size(); ++i) {
-    ASSERT_EQ(pos, strings_->value_offset(i));
-    ASSERT_EQ(static_cast<int>(expected_[i].size()), strings_->value_length(i));
-    pos += static_cast<int>(expected_[i].size());
-  }
-}
-
-TEST_F(TestStringArray, TestDestructor) {
-  auto arr = std::make_shared<StringArray>(length_, offsets_buf_, value_buf_,
+    strings_ = std::make_shared<ArrayType>(length_, offsets_buf_, value_buf_,
                                            null_bitmap_, null_count_);
-}
+  }
 
-TEST_F(TestStringArray, TestGetString) {
-  for (size_t i = 0; i < expected_.size(); ++i) {
-    if (valid_bytes_[i] == 0) {
-      ASSERT_TRUE(strings_->IsNull(i));
+  void _TestArrayBasics() {
+    ASSERT_EQ(length_, strings_->length());
+    ASSERT_EQ(1, strings_->null_count());
+    ASSERT_OK(ValidateArray(*strings_));
+    TestInitialized(*strings_);
+    AssertZeroPadded(*strings_);
+  }
+
+  void _TestType() {
+    std::shared_ptr<DataType> type = this->strings_->type();
```

Comment:
I don't think you need the if/else chain:
```suggestion
    ASSERT_EQ(this->strings_->type()->id(), TypeClass::type_id);
    ASSERT_TRUE(this->strings_->type()->Equals(TypeTraits<TypeClass>::type_singleton()));
```

Reply:
OTOH it makes sure we got the constants right. I'd rather keep this single test as-is.